### ISSUE TITLE
drop meteor does not work via the rest API

### ISSUE THREAD

**Issue by kinvaris** [NyxStudios/TShock#1361]
During a drop meteor call I got this error:
```
{
  "status": "500",
  "error": "Internal server error.",
  "errormsg": "Object reference not set to an instance of an object",
  "stacktrace": "  at Terraria.WorldGen.dropMeteor () [0x00000] in <filename unknown>:0 \n  at TShockAPI.RestManager.WorldMeteor (Rests.RestRequestArgs args) [0x00000] in <filename unknown>:0 \n  at Rests.SecureRestCommand.Execute (Rests.RestVerbs verbs, IParameterCollection parameters, TokenData tokenData, IRequest request, IHttpContext context) [0x00000] in <filename unknown>:0 \n  at Rests.SecureRest.ExecuteCommand (Rests.RestCommand cmd, Rests.RestVerbs verbs, IParameterCollection parms, IRequest request, IHttpContext context) [0x00000] in <filename unknown>:0 \n  at Rests.Rest.ProcessRequest (System.Object sender, HttpServer.RequestEventArgs e) [0x00000] in <filename unknown>:0 "
}
```

Even after generating a new world I got the same error.

**Comment by chainsol** [comment]
What does Terraria.WorldGen.spawnMeteor = False do? It's the only difference I find between the standard server command and the REST command.

**Comment by kinvaris** [comment]
What I should mention is when you should try to reproduce the issue you should have the following OS:
```
terraria@jonas-vps2-venus:~/tshock$ cat /etc/os-release 
PRETTY_NAME="Debian GNU/Linux 8 (jessie)"
NAME="Debian GNU/Linux"
VERSION_ID="8"
VERSION="8 (jessie)"

terraria@jonas-vps2-venus:~/tshock$ uname -a
Linux jonas-vps2-venus 3.16.0-4-amd64 #1 SMP Debian 3.16.7-ckt11-1+deb8u5 (2015-10-09) x86_64 GNU/Linux
```

@chainsol: I am currently running this version: `TShock 4.3.21.0 (Mintaka)`

**Comment by chainsol** [comment]
I'm on TShock: 4.3.21.0 (Mintaka)

As for the OS
PRETTY_NAME="Debian GNU/Linux 8 (jessie)"
NAME="Debian GNU/Linux"
VERSION_ID="8"
VERSION="8 (jessie)"
ID=debian
HOME_URL="http://www.debian.org/"
SUPPORT_URL="http://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"

I can't reproduce your issue. :[

**Comment by kinvaris** [comment]
@chainsol I was able to reproduce the issue.
So what I did was: 

* Start the server (no players were joined)
* Requested a new token: ``
* Spawned the meteor: `Meteor successfully spawned`
* Tried to spawn it a second time: `Meteor successfully spawned`
* Joined a linux terraria client on the server
* Tried to spawn the meteor again:
```
{
  "status": "500",
  "error": "Internal server error.",
  "errormsg": "Object reference not set to an instance of an object",
  "stacktrace": "  at Terraria.WorldGen.dropMeteor () [0x00000] in <filename unknown>:0 \n  at TShockAPI.RestManager.WorldMeteor (Rests.RestRequestArgs args) [0x00000] in <filename unknown>:0 \n  at Rests.SecureRestCommand.Execute (Rests.RestVerbs verbs, IParameterCollection parameters, TokenData tokenData, IRequest request, IHttpContext context) [0x00000] in <filename unknown>:0 \n  at Rests.SecureRest.ExecuteCommand (Rests.RestCommand cmd, Rests.RestVerbs verbs, IParameterCollection parms, IRequest request, IHttpContext context) [0x00000] in <filename unknown>:0 \n  at Rests.Rest.ProcessRequest (System.Object sender, HttpServer.RequestEventArgs e) [0x00000] in <filename unknown>:0 "
}
```

* Disconnected the linux terraria client:
* Tried to spawn the meteor again:
```
{
  "status": "200",
  "response": "Meteor has been spawned"
}
```

**Comment by chainsol** [comment]
Ok, interesting, haha. We'll see if anyone official chimes in - What happens if you run the command /dropmeteor as the logged in user on the server?

**Comment by kinvaris** [comment]
When the `/dropmeteor` is executed from the server (so from the server command terminal). It works. When executing from the REST API via the REST call `/v3/server/rawcmd` the server responds with a success!

**Comment by kinvaris** [comment]
When the `/dropmeteor` is executed from the server (so from the server command terminal). It works.

**Comment by chainsol** [comment]
Ok, here's another doozy - I can't replicate even with a player on the server. I still get a meteor spawned.

**Comment by OnsenManju** [comment]
Same here, but in my case, this happened with the latest TShock (4.3.22) on Ubuntu and Windows 7.
It appears that this depends on world files, so this might happen on any platforms.

As far as I tested, this REST error occurred in the world that I have never executed any of commands as follows:
- `/dropmeteor` in-game
- `/dropmeteor` on console while player(s) is joined

Once the meteor has dropped by the command, the REST error doesn't occur in that world whether someone has connected or not.
(Is there a flag something like "_IsMeteorDropped_" in the world file?)

**Comment by OnsenManju** [comment]
**BUT:** Even if `/world/meteor` endpoint showed no error, any meteor hadn't dropped on the ground in reality.
And also, executing `/dropmeteor` on console without connected players is the same result (even if `-forceupdate` option is used at launching TShock).

It seems that only this message is an evidence for the actual meteor falling.
![image](https://cloud.githubusercontent.com/assets/16373401/21818163/23fcc24e-d7aa-11e6-8516-e516271e8751.png)

**Comment by OnsenManju** [comment]
**Update**
- Once a meteor has dropped, `/dropmeteor` command (in-game & on console) won't work until meteorite blocks are cleared or perhaps certain conditions will be met ([this?](http://terraria.gamepedia.com/Meteorite_(biome)#Conditions))
(Only this will be displayed: "_XXXX trigerred a meteor._")
- After that condition is met (=become able to drop meteor), the REST error occurs again
- Doing `/dropmeteor` command by using `/v3/server/rawcmd` endpoint is the similar result as `/world/meteor`

**Comment by nicatronTg** [comment]
Does this still happen in #1445?

**Comment by OnsenManju** [comment]
Unfortunately it seems not changed for this issue on #1445.

On a fresh world, `http://127.0.0.1:7878/world/meteor?token=tokentext` shows
```
{
  "status": "500",
  "error": "Internal server error.",
  "errormsg": "Object reference not set to an instance of an object.",
  "stacktrace": "
   at Terraria.WorldGen.dropMeteor()\r\n
   at TShockAPI.RestManager.WorldMeteor(RestRequestArgs args)\r\n
   at Rests.SecureRestCommand.Execute(RestVerbs verbs, IParameterCollection parameters, TokenData tokenData, IRequest request, IHttpContext context)\r\n
   at Rests.SecureRest.ExecuteCommand(RestCommand cmd, RestVerbs verbs, IParameterCollection parms, IRequest request, IHttpContext context)\r\n
   at Rests.Rest.ProcessRequest(Object sender, RequestEventArgs e)"
}
```
(newline inserted)

`http://127.0.0.1:7878/v3/server/rawcmd?cmd=/dropmeteor&token=tokentext` shows as follows at browser:
```
{
  "status": "200",
  "response": [
    "Command failed, check logs for more details."
  ]
}
```
and the log file shows
```
Command: ERROR: System.NullReferenceException: Object reference not set to an instance of an object.
   at Terraria.WorldGen.dropMeteor()
   at TShockAPI.Commands.DropMeteor(CommandArgs args)
   at TShockAPI.Command.Run(String msg, Boolean silent, TSPlayer ply, List`1 parms)
```

After I dropped a meteor by using `/dropmeteor` on console,
`/world/meteor` endpoint shows
```
{
  "status": "200",
  "response": "Meteor has been spawned"
}
```
and `/v3/server/rawcmd` endpoint shows
```
{
  "status": "200",
  "response": []
}
```
But both of them didn't drop meteors in reality (because the world is not ready for it).

And, after clearing meteorite blocks, the errors occur again.

**Comment by WhiteXZ** [comment]
@OnsenManju would you be able to try [Travis #1446](https://travis.tshock.co/general-devel/1446/)?
Thanks :)

**Comment by OnsenManju** [comment]
It seems resolved in #1446!

If the world is ready for meteor falling, both `/world/meteor` and `/v3/server/rawcmd?cmd=/dropmeteor` worked successfully (and a meteor has dropped in reality).

via `/world/meteor` endpoint:
```
{
  "status": "200",
  "response": "Meteor has been spawned"
}
```
![image](https://cloud.githubusercontent.com/assets/16373401/21844529/4c74a2cc-d832-11e6-9df2-186b9ef43b96.png)
(Nothing is displayed on console)

via `/v3/server/rawcmd?cmd=/dropmeteor` endpoint:
```
{
  "status": "200",
  "response": []
}
```
![image](https://cloud.githubusercontent.com/assets/16373401/21844545/57ca29f8-d832-11e6-8d6e-41f0c2bed8c0.png)
(It shows "_XXXX(superadmin's name) executed: /dropmeteor._" on console)

Thanks @WhiteXZ and @nicatronTg!

**Comment by WhiteXZ** [comment]
Tyvm for testing this for us :)
